src/Features/CSharp/Portable/Copilot/CSharpCopilotCodeFixProvider.cs

```diff
@@ -0,0 +1,111 @@
+        if (!isAvailable)
+            return;
+
+        var promptTitles = await copilotService.GetAvailablePromptTitlesAsync(document, cancellationToken).ConfigureAwait(false);
+        if (promptTitles.IsEmpty)
+            return;
+
+        var hasMultiplePrompts = promptTitles.Length > 1;
+
+        // Find the containing method, if any, and also update the fix span to the entire method.
+        // TODO: count location in doc-comment as part of the method.
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        var containingMethod = CSharpSyntaxFacts.Instance.GetContainingMethodDeclaration(root, context.Span.Start, useFullSpan: false);
+        if (containingMethod is not BaseMethodDeclarationSyntax)
+            return;
+
+        foreach (var diagnostic in context.Diagnostics)
+        {
+            Debug.Assert(containingMethod.FullSpan.IntersectsWith(diagnostic.Location.SourceSpan));
+
+            var fix = TryGetFix(document, containingMethod, diagnostic, hasMultiplePrompts);
+            if (fix != null)
+                context.RegisterCodeFix(fix, diagnostic);
+        }
+    }
+
+    private static CodeAction? TryGetFix(
+        Document document,
+        SyntaxNode method,
+        Diagnostic diagnostic,
+        bool hasMultiplePrompts)
+    {
+        if (diagnostic.Properties.TryGetValue(FixPropertyName, out var fix) is false ||
+            diagnostic.Properties.TryGetValue(PromptTitlePropertyName, out var promptTitle) is false)
+        {
+            return null;
+        }
+
+        var title = hasMultiplePrompts
+            ? $"{CSharpFeaturesResources.Apply_fix_from} {promptTitle}"
+            : CSharpFeaturesResources.Apply_Copilot_suggestion;
+
+        return new CopilotDocumentChangeCodeAction(
+            title,
+            createChangedDocument: (_, cancellationToken) => GetFixedDocumentAsync(method, fix!, cancellationToken),
+            equivalenceKey: title,
+            new CopilotDismissChangesCodeAction(method, diagnostic),
+            priority: CodeActionPriority.Low);
+
+        async Task<Document> GetFixedDocumentAsync(SyntaxNode method, string fix, CancellationToken cancellationToken)
+        {
+            var editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+            var fixMethod = SyntaxFactory.ParseMemberDeclaration(fix, options: editor.SemanticModel.SyntaxTree.Options)!;
+            var newMethod = fixMethod
+                                .WithLeadingTrivia(fixMethod.HasLeadingTrivia ? fixMethod.GetLeadingTrivia() : method.GetLeadingTrivia())
+                                .WithTrailingTrivia(fixMethod.HasTrailingTrivia ? fixMethod.GetTrailingTrivia() : method.GetTrailingTrivia())
+                                .WithAdditionalAnnotations(Formatter.Annotation, WarningAnnotation);
```

Comment:
is this generating new code from some external source?  if so, replace all whitespace trivia on it with elastic trivia.  That way we get the formatter to fix up everything.  

this all needs docs.